cpp/include/raft/core/mdarray.hpp

```diff
@@ -642,9 +792,146 @@ auto make_device_vector(size_t n, rmm::cuda_stream_view stream)
+auto flatten(const array_interface_type& mda)
+{
+  return flatten(mda.view());
+}
+
+template <typename ElementType, typename LayoutType>
+constexpr auto flatten(host_vector_view<ElementType, LayoutType> h_vv)
+{
+  return h_vv;
+}
+
+template <typename ElementType, typename LayoutType>
+auto flatten(const host_vector<ElementType, LayoutType>& h_v)
+{
+  return flatten(h_v.view());
+}
+
+template <typename ElementType>
+constexpr auto flatten(host_scalar_view<ElementType> h_sv)
+{
+  return h_sv;
+}
+
+template <typename ElementType>
+auto flatten(const host_scalar<ElementType>& h_s)
+{
+  return flatten(h_s.view());
+}
+
+template <typename ElementType, typename LayoutType>
+constexpr auto flatten(device_vector_view<ElementType, LayoutType> d_vv)
+{
+  return d_vv;
+}
+
+template <typename ElementType, typename LayoutType>
+auto flatten(const device_vector<ElementType, LayoutType>& d_v)
+{
+  return flatten(d_v.view());
+}
+
+template <typename ElementType>
+constexpr auto flatten(device_scalar_view<ElementType> d_sv)
+{
+  return d_sv;
+}
+
+template <typename ElementType>
+auto flatten(const device_scalar<ElementType>& d_s)
+{
+  return flatten(d_s.view());
+}
+
+/**
+ * @brief Reshape raft::host_mdspan or raft::device_mdspan
+ *
+ * @tparam mdspan_type Expected type raft::host_mdspan or raft::device_mdspan
+ * @tparam Extents raft::extents for dimensions
+ * @param mds raft::host_mdspan or raft::device_mdspan object
+ * @param new_shape Desired new shape of the input
+ * @return raft::host_mdspan or raft::device_mdspan, depending on AccessorPolicy
+ */
+template <typename mdspan_type,
+          size_t... Extents,
+          std::enable_if_t<is_mdspan_v<mdspan_type>>* = nullptr>
+auto reshape(mdspan_type mds, extents<Extents...> new_shape)
+{
+  RAFT_EXPECTS(mds.is_contiguous(), "Input must be contiguous.");
+
+  size_t new_size = 1;
+  for (size_t i = 0; i < new_shape.rank(); ++i) {
+    new_size *= new_shape.extent(i);
+  }
+  RAFT_EXPECTS(new_size == mds.size(), "Cannot reshape array with size mismatch");
+
+  return detail::stdex::mdspan<typename mdspan_type::element_type,
+                               decltype(new_shape),
+                               typename mdspan_type::layout_type,
+                               typename mdspan_type::accessor_type>(mds.data(), new_shape);
+}
+
+/**
+ * @brief Reshape object implementing raft::array_interface
+ *
+ * @tparam array_interface_type Expected type implementing raft::array_interface
+ * @tparam Extents raft::extents for dimensions
+ * @param mda raft::array_interace implementing object
+ * @param new_shape Desired new shape of the input
+ * @return raft::host_mdspan or raft::device_mdspan, depending on the underlying
+ *         ContainerPolicy
+ */
+template <typename array_interface_type,
+          size_t... Extents,
+          std::enable_if_t<is_array_interface_v<array_interface_type>>* = nullptr>
+auto reshape(const array_interface_type& mda, extents<Extents...> new_shape)
```

Comment:
I haven't looked into the array interface in raft yet.  But if this function accepts (or will potentially accept in the future) an owning type, and might change the size of the underlying memory buffer, one might want to accept a raft handle as a parameter.